### ISSUE TITLE
Built wheel is invalid: Metadata 1.2 mandates PEP 440 version, but '0.0.2.dev1.post1.dev2-g8fd2842' is not

### ISSUE THREAD

**Issue by JulianOrteil** [pypa/setuptools-scm#654]
Attempting to run tox tests on a library I am building is throwing the below error:

```
Processing ./.tox/.tmp/package/1/python-step-series-0.0.2.dev1.post1.dev2-g8fd2842.tar.gz
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
    Preparing wheel metadata: started
    Preparing wheel metadata: finished with status 'done'
Building wheels for collected packages: python-step-series
  Building wheel for python-step-series (PEP 517): started
  Building wheel for python-step-series (PEP 517): finished with status 'done'
  Created wheel for python-step-series: filename=python_step_series-0.0.2.dev1.post1.dev2_g8fd2842-py3-none-any.whl size=18021 sha256=7c172e5d554075b0c12255e06be171c067edf4569bb621c5fa573070d39d581d
  Stored in directory: /home/user/.cache/pip/wheels/d1/5f/e0/84fd855c772a59469c7065c94464233a7532667a9c94430803
  WARNING: Built wheel for python-step-series is invalid: Metadata 1.2 mandates PEP 440 version, but '0.0.2.dev1.post1.dev2-g8fd2842' is not
Failed to build python-step-series
ERROR: Could not build wheels for python-step-series which use PEP 517 and cannot be installed directly
```

I believe it is being caused by the warning I've titled this issue with.
I just started getting this error this morning and am completely stuck on why I am all of a sudden seeing it. I looked at #278 as it seemed similar, but was not. Regardless, as like in that issue, here are the outputs for `git tag`, `git describe` and `git status`:

`git tag`:
```
0.0.2dev0
0.0.2dev1
v0.0.1
```

`git describe`:
```
0.0.2dev0-6-g8fd2842
```

`git status`:
```
On branch 0.0.2
Your branch is up to date with 'origin/0.0.2'.

nothing to commit, working tree clean
```

Curiously, creating a new tag (i.e. `0.0.2dev2`) will allow me to run the tests until my working tree gets dirty again, so creating a new tag every time I want to test isn't viable nor proper.

I am 100% stuck, so please let me know what I can do to rectify this if you are able.
Jules

**Comment by RonnyPfannschmidt** [comment]
when i run `setup.py --version` and other commands i get valid versions

i only see the versions with the broken dash happen for you, so more detail on your deployments is needed,
in a direct checkout i cant replicate it

this may also be related to pyscaffold (note that the autogenerated run/install instructions are incorrect)

**Comment by RonnyPfannschmidt** [comment]
with both tox3 and tox4 i also cannot replicate the issue

**Comment by JulianOrteil** [comment]
> (note that the autogenerated run/install instructions are incorrect)

Do you mean the CONTRIBUTING file in the project? If so, then we should probably file an issue with pyscaffold directly. I haven't read through it apart from just looking for "todos", so I'm not suprised.

At any rate, let me blow away my environment again and start over. I will let you know if anything changes.

**Comment by JulianOrteil** [comment]
Apparently third time's the charm. Whatever issue git was having was related to just my local repo or to pip's cache since I deleted it as well. I apologize for wasting your time and thank you for your diligence in helping me solve this.

Jules

**Comment by RonnyPfannschmidt** [comment]
@JulianOrteil i'm glad it resolved, bad cache/state based issues are a pain to figure :)

**Comment by JulianOrteil** [comment]
Hi @RonnyPfannschmidt

Apologies for reopening this, I am once again having this issue. Resetting my environment is not working this time but I believe you'll be able to replicate it this time, however. If you checkout branch 0.0.2 in my project (you will need to run a git pull if you already have it cloned) you should see the problem I am having. I've also updated the CONTRIBUTING file (use Miniconda, not virtualenv).

For your convenience, here are the outputs of `git status`, `git describe`,  `git tag` and the contents of my `pyproject.toml`:

`git status`:
```
On branch 0.0.2
Your branch is up to date with 'origin/0.0.2'.

nothing to commit, working tree clean
```

`git describe`:
```
v0.0.2dev2-1-gcaf9278
```

`git tag`:
```
v0.0.1
v0.0.2dev0
v0.0.2dev1
v0.0.2dev2
```

`pyproject.toml`:
```
[build-system]
# AVOID CHANGING REQUIRES: IT WILL BE UPDATED BY PYSCAFFOLD!
requires = ["setuptools>=46.1.0", "setuptools_scm[toml]>=5", "wheel"]
build-backend = "setuptools.build_meta"

[tool.setuptools_scm]
# See configuration details in https://github.com/pypa/setuptools_scm
version_scheme = "no-guess-dev"
```

Thanks again for your time.
Jules

**Comment by JulianOrteil** [comment]
Ping

**Comment by RonnyPfannschmidt** [comment]
Apologies, im a bit tied up with personal matters, i hope to finally get a look at this today

**Comment by RonnyPfannschmidt** [comment]
i suspect this is related to bad tags, i may have to make setuptools_scm error out on those

i can replicate the generated extra versions, but the absolutely and inherently wrong `-` i cannot replicate , please provide details on the commands you use for building

**Comment by RonnyPfannschmidt** [comment]
@JulianOrteil i can now rel´plcate the issue with tox, i think i need the help of @gaborbernat to figure what exactly breaks here

**Comment by RonnyPfannschmidt** [comment]
i identified a normalization issue that sticks into a sdist

**Comment by RonnyPfannschmidt** [comment]
@JulianOrteil i determined that your use of dev tags is wrong for the choosen version scheme and results in version number that setuptools will normalize wrongly

setuptools_scm will be made to fail with such tags in the next version,

i will have to figure a alternative version scheme

@FlorianWilhelm why did pyscaffold add a default to no-guess-dev

**Comment by JulianOrteil** [comment]
Hi Ron, I hope everything clears up for you personally. Thank you for looking into my issue more.

I would not have suspected the dev part of my tags to be the issue so I'm grateful for your persistence. Let me know if there is an alternative you would prefer me to use. I'll be sure to use that from now on.

Thanks again,
Jules

**Comment by FlorianWilhelm** [comment]
@RonnyPfannschmidt We default to `no-guess-dev` since for me it's the only sane default. Can you really ever guess the next version right? After version 1.3.2 could come 1.3.3 or 1.4 or even 2.0, right? So `no-guess-dev` feels more deterministic using the .post1.devN trick. CC @abravalheri

**Comment by RonnyPfannschmidt** [comment]
@FlorianWilhelm that's why one can do dev0 tags and set up the next line

**Comment by RonnyPfannschmidt** [comment]
@FlorianWilhelm i guess i need to document the details better, but no gues dev is a last resort, not a good default

**Comment by FlorianWilhelm** [comment]
@RonnyPfannschmidt Not sure I understand the concept of dev0 tags, does this mean I need to first manually add a git tag like `v1.3.dev0` and then setuptools_scm knows that the next version to guess is `v1.4`? If this is how it is supposed to work I think it's not really user-friendly. I am already having a hard time motivating aspiring data scientists to follow semantic versioning. If they would also need to anticipate if they next version is major, or minor, or bugfix (things change and not everyone uses complicated git workflows like git flow) I think this would raise the acceptance hurdle even further.

Besides the user's perspective, what if you want to build every commit and push it into a testing index on devpi? This worked quite well with the `post1.dev`-trick but I couldn't image to set this up robustly when guessing the next version.

**Comment by RonnyPfannschmidt** [comment]
Calver seems to be a fit for those situations

For the rest please open a new issue so that we can flesh out the use case

**Comment by FlorianWilhelm** [comment]
I am fine with what no-guess-dev does. How about a call coming week? I think it's easier to just discuss it over phone.

**Comment by RonnyPfannschmidt** [comment]
I absolutely agree that a call can massively speed up the discussion, I still want a extra issue to record the details /conclusion

**Comment by RonnyPfannschmidt** [comment]
the next release will error out on those invalid tags

**Comment by RonnyPfannschmidt** [comment]
resolved with the 6.4.0 release
